src/esw/can_node/CanNode.hpp

```diff
@@ -0,0 +1,146 @@
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+#include <cstdio>
+#include <cstring>
+#include <format>
+#include <iostream>
+#include <linux/can.h>
+#include <linux/can/raw.h>
+#include <net/if.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <vector>
+
+class CanNode {
+public:
+    CanNode(bool _isExtendedFrame) : isExtendedFrame(_isExtendedFrame) {
+        sockaddr_can addr{};
+        ifreq ifr{};
+
+        if ((s = socket(PF_CAN, SOCK_RAW, CAN_RAW)) < 0) {
+            perror("Error while opening socket");
+        }
+
+        const char* ifname = "can0";
+        strcpy(ifr.ifr_name, ifname);
+        ioctl(s, SIOCGIFINDEX, &ifr);
+
+        addr.can_family = AF_CAN;
+        addr.can_ifindex = ifr.ifr_ifindex;
+
+        if (bind(s, (struct sockaddr*) &addr, sizeof(addr)) < 0) {
+            perror("Error in socket bind");
+        }
+    }
+
+    void send_frame() const {
+        // write(s, &frame, sizeof(struct can_frame));
+        sendto(s, &frame, sizeof(struct can_frame), 0, nullptr, 0);
+    }
+
+    void set_bus(uint8_t bus) {
+        this->bus = bus;
+    }
+
+    // canfd_frame.can_id is a uint32_t with format:
+    // [0-28]: CAN identifier (11/29bit)
+    // [29]: Error frame flag (0 = data frame, 1 = error frame)
+    // [30]: Remote transmission request flag (1 = rtr frame)
+    // [31]: Frame format flag (0 = standard 11bit, 1 = extended 29bit)
+    // In the future, if we want to send different types of messages,
+    // we should have logic for switching bits such as errorFrameFlag.
+    void set_frame_id(uint32_t id) {
+        uint32_t frameFormatFlag;
+        uint32_t identifier;
+        if (isExtendedFrame) {
+            frameFormatFlag = 0x80000000; // set can_id[31] high for extended frame
+            identifier = id & 0x01FFFFFF; // lower 29 bit mask of id (extended)
+        } else {
+            frameFormatFlag = 0x00000000; // set can_id[31] low for standard frame
+            identifier = id & 0x000007FF; // lower 11 bit mask of id (standard)
+        }
+        uint32_t errorFrameFlag = 0x20000000; // set can_id[29] high
+        uint32_t rtrFlag = 0x40000000;        // set can_id[30] high
+
+        frame.can_id = identifier | errorFrameFlag | rtrFlag | frameFormatFlag;
+    }
+
+    void set_frame_data(std::span<std::byte> const& data) {
+        for (size_t i = 0; i < data.size(); ++i) {
```

Comment:
This loop can just be a `std::memcpy`

Reply:
`data` will give you access to `.data()` which the raw pointer and `.size()` which is the length in bytes